lib/tasks/demo.rake

```diff
@@ -36,17 +36,49 @@ namespace :demo do
   desc 'Generate N many more monitorees based on existing data'
   task create_bulk_data: :environment do
     raise 'This task is only for use in a development environment' unless Rails.env == 'development' || ENV['DISABLE_DATABASE_ENVIRONMENT_CHECK']
+
     num_patients = (ENV['COUNT'] || 100000).to_i
-    num_threads = (ENV['THREADS'] || 8).to_i
+    num_forks = (ENV['FORKS'] || 8).to_i
 
-    duplicateable = Patient.where('responder_id = id').pluck(:id)
-    threads = []
-    (1..num_threads).each do |x|
-      threads << Thread.new do  (1..(num_patients/num_threads)).each {|x| deep_duplicate_patient(Patient.find(duplicateable.sample)) } end
-    end
+    # Found that forks were consistently performing drastically differently when id's were not shuffled
+    # i.e. fork 1 was always the slowest between 5 to 10 p/s and fork 8 was always the fastest between 20 to 25 p/s
+    # After shuffling, all forks end up at a much more similar 9 to 10 p/s
+    patient_ids = Patient.where('patients.responder_id = patients.id').limit(num_patients).pluck(:id).shuffle
+
+    pids = []
+
+    ::ActiveRecord::Base.clear_all_connections!
+    fork_num = 1
+    patient_ids.each_slice(patient_ids.size / num_forks).each do |slice_ids|
+      if fork_num > 8
+        puts 'Hit maximum fork size. Will not create new fork.'
+        break
+      end
+
+      pids << fork do
+        t1 = Time.now
+        ::ActiveRecord::Base.establish_connection
 
-    threads.each(&:join)
+        num_to_create = num_patients / num_forks
+        num_created = 0
 
+        while num_created < num_to_create do
+          # deep_duplicate returns exactly how many patients were created (including duplicated dependents)
+          num_created += deep_duplicate_patient(Patient.find(slice_ids.sample))
+          print "\r#{(num_created / (Time.now - t1)).truncate(2)} p/s"
```

Comment:
can we also print `num_created` along with the rate so it's easy to get an idea of how many patients have been duped so far?

Reply:
That would be really nice if it was total patients over all forks, but since num_created is per-fork, this would end up writing very different values that get further apart as the task runs and not be super useful IMO.

I tried previously to figure out a way to print out one line for each fork that is overwritten with each fork's iteration

```
fork 1: 1,423 patients created @ 8.12 p/s
fork 2: 1,553 patients created @ 9.24 p/s
...
```

If anyone has any ideas on how to accomplish this then I'd be happy to implement

Reply:
did it here: https://github.com/SaraAlert/SaraAlert/pull/732/commits/a9754877f63eb4728d8202f5d2ee8cba95fa80b7